File: MobileApp/my_social_app/lib/models/transaction.dart
import 'package:flutter/material.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:multimedia/models/multimedia.dart';
import 'package:my_social_app/state/app_state/transaction_state/transaction_state.dart';
part 'transaction.g.dart';

@immutable
@JsonSerializable()
class Transaction {
  final int id;
  final DateTime createdAt;
  final int balanceId;
  final int numberOfInputToken;
  final int numberOfOutputToken;
  final int solPerInputToken;
  final int solPerOutputToken;
  final String aiModelName;
  final Multimedia aiModelImage;

  const Transaction({
    required this.id,
    required this.createdAt,
    required this.balanceId,
    required this.numberOfInputToken,
    required this.numberOfOutputToken,
    required this.solPerInputToken,
    required this.solPerOutputToken,
    required this.aiModelName,
    required this.aiModelImage,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) => _$TransactionFromJson(json);
  Map<String, dynamic> toJson() => _$TransactionToJson(this);


  TransactionState toState() =>
    TransactionState(
      id: id,
      createdAt: createdAt,
      balanceId: balanceId,
      numberOfInputToken: numberOfInputToken,
      numberOfOutputToken: numberOfOutputToken,
      solPerInputToken: solPerInputToken,
      solPerOutputToken: solPerOutputToken,
      aiModelName: aiModelName,
      aiModelImage: aiModelImage
    );
}

File: src/SolTake.Domain/TransactionAggregate/Entities/Transaction.cs
using SolTake.Core;

namespace SolTake.Domain.TransactionAggregate.Entities
{
    public class Transaction : Entity, IAggregateRoot
    {
        public int BalanceId { get; private set; }
        public int NumberOfSol { get; private set; }

        public Transaction(int balanceId, int numberOfInputToken, int numberOfOutputToken, int solPerInputToken, int solPerOutputToken)
        {
            BalanceId = balanceId;
            NumberOfSol = -1 * numberOfInputToken * solPerInputToken + numberOfOutputToken * solPerOutputToken;
        }

        public Transaction(int balanceId, int numberOfSol)
        {
            BalanceId = balanceId;
            NumberOfSol = numberOfSol;
        }
        
        private Transaction() { }
        
        public void Create() => CreatedAt = DateTime.UtcNow;
    }
}
